enhancements/workload-partitioning/rationalizing-configuration.md

```diff
@@ -0,0 +1,213 @@
+
+The current implementation is prone to user error, because it has two
+entities configuring and managing the same data. This design document
+describes phase 2 of the implementation plan for workload partitioning
+to improve the user experience when configuring the feature.
+
+### Goals
+
+1. Describe the remaining work at a high level for architectural
+   review and planning.
+2. Document the division of responsibility between workload management
+   in components in the OpenShift release payload and PAO.
+3. List the other enhancements that need to be written, and their
+   scopes.
+
+### Non-Goals
+
+1. Resolve all of the details for each of the next steps. We're not
+   going to talk a lot about names or implementation details here.
+
+## Proposal
+
+### API
+
+The initial implementation is limited to single-node deployments. To
+support multi-node clusters, we need to solve a couple of concurrency
+problems. The first is providing a way for the admission hook
+responsible for mutating pod definitions to reliably know when the
+feature should be enabled. The admission hook cannot store state on
+its own, so we need a new API to indicate when the feature is
+enabled.
+
+We eventually want to extend the partitioning feature to support
+non-management workloads to include workload types defined by cluster
+admins. For now, however, we want to keep the interface as simple as
+possible.  Therefore, the new API will be read-only, and only used so
+that the admission hook (and other consumers) can tell when workload
+partitioning is active.
+
+The new API will need a controller to manage it. We propose to add a
+controller to the existing cluster-config-operator, to avoid adding
+the overhead of another image, pod, etc. to the cluster.
+
+We propose that the API be owned by core OpenShift, and defined in the
+`openshift/api` repository. A separate enhancement will be written to
+describe the API and controller in more detail.
+
+### Owner for Enablement
+
+Workload management needs two node configuration steps which overlap
+with work PAO is doing today:
+
+* Set reserved CPUs in kubelet config and enable the CPU manager
+  static policy
+* Set systemd affinity to match the reserved CPU set
+
+PAO already has logic for managing CPU pools and configuring kubelet
+and CRI-O. Since it already does a lot of what is needed to enable
+workload partitioning, we propose to extend it to handle the
+additional configuration currently being done manually by users,
+including the machine config with settings for kubelet and CRI-O.
+
+The effect of this decision is that the easiest way to use the
+workload partitioning feature will be through the
+performance-addon-operator. This means that the implementation of the
+feature is split across a few components, and not all of them are
+delivered in all clusters, but it avoids duplicating a lot of the work
+already done in the PAO.
+
+Workload partitioning is currently enabled as part of installing a
```

Comment:
so is the flow here
1. user expresses desire for management CPU pool
2. PAO configures the kubeletconfig and systemd somehow
3. PAO watches the state of kubelet config and systemd and updates an API to indicate to the admission plugin: "this is now enabled"

Reply:
Not quite.

We thought step 3 should not rely on PAO, because the admission plugin always needs to be able to answer the question about whether partitioning is enabled for a particular type. We also don't want to tie that enablement to the implementation details within kubelet or CRI-O. Instead, either the cluster-config-operator or machine-config-operator will have a new controller that looks at the nodes in the cluster to track when they advertise the partition-specific opaque resources (caused by the kubelet config being updated) and manage a status API (to be designed in another enhancement).

Reply:
> We thought step 3 should not rely on PAO, because the admission plugin always needs to be able to answer the question about whether partitioning is enabled for a particular type.

Above you indicated that the "PerformanceProfile CRD" is used to tell the PAO to configure this.  If the flow requires the PAO to enable, then requiring the PAO to set your status indicator CRD seems reasonable.  The admission plugin could still be active all the time, but at least the user desire and the state of the cluster would be managed by a single component.

Reply:
So, the status API would be defined in the `openshift/api` repo and available to both the admission plugin and PAO. To keep the admission hook from having a dependency on something defined in PAO, that API needs to not be the PerformanceProfile itself. The CRD for the status API would always be installed, but there may not always be an instance. Something in PAO would create the instance and keep it up to date.

Does that make sense as an approach?

Reply:
Creating another object is easy enough, so yes, PAO can do this. The only question is whether it should be always enabled or whether we should create the instance based on something (master+worker case vs. SNO vs. RWN vs. three nodes total etc..).

Reply:
> The CRD for the status API would always be installed, but there may not always be an instance. Something in PAO would create the instance and keep it up to date.
> 
> Does that make sense as an approach?

That could work.  So could attaching a status field to an existing resource (is this related to Infrastructure?) that is empty until the PAO sets the value.

Reply:
We have updated the text to always use the PAO to configure partitioning.